ci/tasks.py

```diff
@@ -0,0 +1,259 @@
+
+    # we could consider labeling secrets and pulling all secrets by label
+    # but that moves some logic into Secret Manager itself
+    secrets: List[str] = []
+
+    # for helm
+    namespace: Optional[str]
+    helm_name: Optional[str]
+    helm_chart: Optional[Path]
+    helm_values: List[Path] = []
+
+    # for kustomize
+    kustomize_dir: Optional[Path]
+
+
+# TODO: rename this
+class Config(BaseModel):
+    git_repo: git.Repo
+    channel: str  # this is a bit weird, but I want to be able to log this value
+    releases: List[Release]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    @validator("git_repo", pre=True)
+    def parse_git_repo(cls, v):
+        return git.Repo(v, search_parent_directories=True)
+
+    @property
+    def git_root(self) -> Optional[Path]:
+        if not self.git_repo.working_tree_dir:
+            return None
+        return Path(self.git_repo.working_tree_dir)
+
+
+@task
+def parse_calitp_config(c):
+    """
+    Parses the top-level calitp configuration key via Pydantic
+    """
+    c.update({"calitp_config": Config(**c.config._config["calitp"])})
+
+
+def get_releases(
+    c,
+    driver: Optional[ReleaseDriver] = None,
+    app=None,
+) -> List[Release]:
+    ret = []
+    for release in c.calitp_config.releases:
+        if (not driver or release.driver == driver) and (
+            not app or app == release.name
+        ):
+            ret.append(release)
+    return ret
+
+
+# TODO: kubectl diff now supports masking secrets, so we shoulds be able to render secret diffs in PRs
+#  see https://github.com/kubernetes/kubernetes/pull/96084
+@task(
+    parse_calitp_config,
+    help={
+        "app": GENERIC_HELP["app"],
+        "secret": "Optionally, specify a single secret to deploy",
+    },
+)
+def secrets(
+    c,
+    app=None,
+    secret=None,
+    hide=True,
+):
+    """
+    Deploy secret(s) by channel, and optionally app or secret name.
+    """
+    client = secretmanager.SecretManagerServiceClient()
+    found_secret = False
+    for release in get_releases(c, app=app):
+        for release_secret in release.secrets:
+            if not secret or secret == release_secret:
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    secret_path = Path(tmpdir) / Path(f"{release_secret}.yml")
+                    # this ID maps to cal-itp-data-infra; there's probably a better way to do this
+                    # TODO: we could probably just use gcloud CLI for this
+                    name = f"projects/1005246706141/secrets/{release_secret}/versions/latest"
+                    secret_contents = client.access_secret_version(
+                        request={"name": name}
+                    ).payload.data.decode("UTF-8")
+
+                    if release.namespace:
+                        ns_str = f"--namespace {release.namespace}"
+                    else:
+                        ns_str = ""
+                        assert (
```

Comment:
not sure if this is off base, but this kind of `assert` with a yaml safe load makes me think that something here could be turned into a pydantic class instead?

Reply:
Not necessarily in this case; a Secret without a namespace is totally valid, but in our specific case it's fine if it is specified in the Release